From 4368438a2233e1fed99eaadd13019bd1ae6fd0f0 Mon Sep 17 00:00:00 2001
From: Chad Dupuis <cdupuis@redhat.com>
Date: Thu, 21 Aug 2014 13:06:40 -0400
Subject: [ethernet] qlcnic: Limit vNIC support in legacy interrupt mode

Message-id: <1408626422-3941-117-git-send-email-cdupuis@redhat.com>
Patchwork-id: 88646
O-Subject: [PATCH RHEL 7.1 116/138] qlcnic: Limit vNIC support in legacy interrupt mode
Bugzilla: 1089352
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

From: Sucheta Chakraborty <sucheta.chakraborty@qlogic.com>

Bug 1089352 (https://bugzilla.redhat.com/show_bug.cgi?id=1089352)

net commit id aaecf51cf31160262b29a6d50f364f4a76c7ed1e

o When the driver loads in legacy interrupt mode, only vNICs
  with PCI function number 0-7 are supported.

Signed-off-by: Sucheta Chakraborty <sucheta.chakraborty@qlogic.com>
Signed-off-by: David S. Miller <davem@davemloft.net>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/net/ethernet/qlogic/qlcnic/qlcnic_83xx_hw.c b/drivers/net/ethernet/qlogic/qlcnic/qlcnic_83xx_hw.c
index 1b5d375..67b70a2 100644
--- a/drivers/net/ethernet/qlogic/qlcnic/qlcnic_83xx_hw.c
+++ b/drivers/net/ethernet/qlogic/qlcnic/qlcnic_83xx_hw.c
@@ -33,6 +33,7 @@ static void qlcnic_83xx_get_beacon_state(struct qlcnic_adapter *);
 #define RSS_HASHTYPE_IP_TCP  0x3
 #define QLC_83XX_FW_MBX_CMD  0
 #define QLC_SKIP_INACTIVE_PCI_REGS 7
+#define QLC_MAX_LEGACY_FUNC_SUPP 8
 
 static const struct qlcnic_mailbox_metadata qlcnic_83xx_mbx_tbl[] = {
  {QLCNIC_CMD_CONFIGURE_IP_ADDR, 6, 1},
@@ -357,8 +358,15 @@ int qlcnic_83xx_setup_intr(struct qlcnic_adapter *adapter)
  if (!ahw->intr_tbl)
   return -ENOMEM;
 
- if (!(adapter->flags & QLCNIC_MSIX_ENABLED))
+ if (!(adapter->flags & QLCNIC_MSIX_ENABLED)) {
+  if (adapter->ahw->pci_func >= QLC_MAX_LEGACY_FUNC_SUPP) {
+   dev_err(&adapter->pdev->dev, "PCI function number 8 and higher are not supported with legacy interrupt, func 0x%x\n",
+    ahw->pci_func);
+   return -EOPNOTSUPP;
+  }
+
   qlcnic_83xx_enable_legacy(adapter);
+ }
 
  for (i = 0; i < num_msix; i++) {
   if (adapter->flags & QLCNIC_MSIX_ENABLED)
diff --git a/drivers/net/ethernet/qlogic/qlcnic/qlcnic_main.c b/drivers/net/ethernet/qlogic/qlcnic/qlcnic_main.c
index 2dbd77c..2ddd0f9 100644
--- a/drivers/net/ethernet/qlogic/qlcnic/qlcnic_main.c
+++ b/drivers/net/ethernet/qlogic/qlcnic/qlcnic_main.c
@@ -2538,6 +2538,9 @@ qlcnic_probe(struct pci_dev *pdev, const struct pci_device_id *ent)
    case -ENOMEM:
     dev_err(&pdev->dev, "Adapter initialization failed. Please reboot\n");
     goto err_out_free_hw;
+   case -EOPNOTSUPP:
+    dev_err(&pdev->dev, "Adapter initialization failed\n");
+    goto err_out_free_hw;
    default:
     dev_err(&pdev->dev, "Adapter initialization failed. Driver will load in maintenance mode to recover the adapter using the application\n");
     goto err_out_maintenance_mode;
-- 
1.7.1